File: Sliding Window/424. Longest Repeating Character Replacement/424. Longest Repeating Character Replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function (s, k) {
  const characters = new Array(26).fill(0);
  let longest = 0;
  let l = 0;

  for (let r = 0; r < s.length; r++) {
    characters[s[r].charCodeAt(0) - "A".charCodeAt(0)]++;

    if (r - l + 1 - Math.max(...characters) > k) {
      characters[s[l].charCodeAt(0) - "A".charCodeAt(0)]--;
      l++;
    }

    longest = Math.max(longest, r - l + 1);
  }

  return longest;
};

// Time: O(n)
// Space: O(1)

File: Sliding Window/424. Longest Repeating Character Replacement/424. Longest Repeating Character Replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        characters = [0] * 26
        longest = 0
        l = 0

        for r in range(len(s)):
            characters[ord(s[r]) - ord("A")] += 1

            while r - l + 1 - max(characters) > k:
                characters[ord(s[l]) - ord("A")] -= 1
                l += 1
            
            longest = max(longest, r - l + 1)

        return longest
        
# Time: O(n)
# Space: O(1)